File: backend/src/main/java/br/ufpr/sept/so2/modules/solicitacoes/application/FormSchemaValidator.java
package br.ufpr.sept.so2.modules.solicitacoes.application;

import br.ufpr.sept.so2.shared.domain.exception.DadoInvalidoException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

@Component
public class FormSchemaValidator {

    private final ObjectMapper objectMapper;

    public FormSchemaValidator(ObjectMapper objectMapper) {
        this.objectMapper = objectMapper;
    }

    public void validar(String formSchemaJson, Map<String, Object> payload) {
        JsonNode schema;
        JsonNode dados;
        try {
            schema = objectMapper.readTree(formSchemaJson);
            dados = objectMapper.valueToTree(payload == null ? Map.of() : payload);
        } catch (Exception ex) {
            throw new DadoInvalidoException("form_schema ou payload inválido.");
        }
        if (!"object".equals(schema.path("type").asText("object"))) {
            throw new DadoInvalidoException("form_schema deve descrever um objeto.");
        }
        JsonNode properties = schema.path("properties");
        if (schema.path("additionalProperties").isBoolean() && !schema.get("additionalProperties").asBoolean()) {
            Set<String> permitidos = new HashSet<>();
            properties.fieldNames().forEachRemaining(permitidos::add);
            dados.fieldNames().forEachRemaining(nome -> {
                if (!permitidos.contains(nome)) {
                    throw new DadoInvalidoException("Campo não previsto no formulário: " + nome);
                }
            });
        }
        List<String> obrigatorios = new ArrayList<>();
        schema.path("required").forEach(node -> obrigatorios.add(node.asText()));
        for (String campo : obrigatorios) {
            JsonNode valor = dados.get(campo);
            if (valor == null || valor.isNull() || (valor.isTextual() && valor.asText().isBlank())) {
                throw new DadoInvalidoException("Campo obrigatório: " + campo);
            }
        }
        properties.fields().forEachRemaining(entry -> {
            JsonNode valor = dados.get(entry.getKey());
            if (valor != null && !valor.isNull()) {
                validarCampo(entry.getKey(), entry.getValue(), valor);
            }
        });
    }

    private static void validarCampo(String nome, JsonNode schema, JsonNode valor) {
        String tipo = schema.path("type").asText("string");
        switch (tipo) {
            case "string" -> {
                if (!valor.isTextual()) {
                    throw new DadoInvalidoException("Campo " + nome + " deve ser texto.");
                }
                String texto = valor.asText();
                if (schema.has("minLength") && texto.length() < schema.get("minLength").asInt()) {
                    throw new DadoInvalidoException("Campo " + nome + " é curto demais.");
                }
                if (schema.has("maxLength") && texto.length() > schema.get("maxLength").asInt()) {
                    throw new DadoInvalidoException("Campo " + nome + " é longo demais.");
                }
            }
            case "number", "integer" -> {
                if (!valor.isNumber()) {
                    throw new DadoInvalidoException("Campo " + nome + " deve ser numérico.");
                }
                if ("integer".equals(tipo) && !valor.isIntegralNumber()) {
                    throw new DadoInvalidoException("Campo " + nome + " deve ser inteiro.");
                }
            }
            case "boolean" -> {
                if (!valor.isBoolean()) {
                    throw new DadoInvalidoException("Campo " + nome + " deve ser verdadeiro ou falso.");
                }
            }
            default -> {
                // tipos compostos ficam para versões futuras do motor
            }
        }
        if (schema.has("enum")) {
            boolean bate = false;
            for (JsonNode opcao : schema.get("enum")) {
                if (opcao.equals(valor)) {
                    bate = true;
                    break;
                }
            }
            if (!bate) {
                throw new DadoInvalidoException("Campo " + nome + " possui valor fora da lista permitida.");
            }
        }
    }
}

File: backend/src/main/kotlin/br/ufpr/sept/so2/modules/solicitacoes/application/FormSchemaValidator.kt
package br.ufpr.sept.so2.modules.solicitacoes.application

import br.ufpr.sept.so2.shared.domain.exception.DadoInvalidoException
import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import org.springframework.stereotype.Component

@Component
class FormSchemaValidator(
    private val objectMapper: ObjectMapper,
) {

    fun validar(formSchemaJson: String?, payload: Map<String, Any?>?) {
        val schema: JsonNode
        val dados: JsonNode
        try {
            schema = objectMapper.readTree(formSchemaJson)
            dados = objectMapper.valueToTree(payload ?: emptyMap<String, Any?>())
        } catch (_: Exception) {
            throw DadoInvalidoException("form_schema ou payload inválido.")
        }
        if ("object" != schema.path("type").asText("object")) {
            throw DadoInvalidoException("form_schema deve descrever um objeto.")
        }
        val properties = schema.path("properties")
        if (schema.path("additionalProperties").isBoolean && !schema.get("additionalProperties").asBoolean()) {
            val permitidos = HashSet<String>()
            properties.fieldNames().forEachRemaining { permitidos.add(it) }
            dados.fieldNames().forEachRemaining { nome ->
                if (!permitidos.contains(nome)) {
                    throw DadoInvalidoException("Campo não previsto no formulário: $nome")
                }
            }
        }
        val obrigatorios = ArrayList<String>()
        schema.path("required").forEach { node -> obrigatorios.add(node.asText()) }
        for (campo in obrigatorios) {
            val valor = dados.get(campo)
            if (valor == null || valor.isNull || (valor.isTextual && valor.asText().isBlank())) {
                throw DadoInvalidoException("Campo obrigatório: $campo")
            }
        }
        properties.fields().forEachRemaining { entry ->
            val valor = dados.get(entry.key)
            if (valor != null && !valor.isNull) {
                validarCampo(entry.key, entry.value, valor)
            }
        }
    }

    companion object {
        private fun validarCampo(nome: String, schema: JsonNode, valor: JsonNode) {
            val tipo = schema.path("type").asText("string")
            when (tipo) {
                "string" -> {
                    if (!valor.isTextual) {
                        throw DadoInvalidoException("Campo $nome deve ser texto.")
                    }
                    val texto = valor.asText()
                    if (schema.has("minLength") && texto.length < schema.get("minLength").asInt()) {
                        throw DadoInvalidoException("Campo $nome é curto demais.")
                    }
                    if (schema.has("maxLength") && texto.length > schema.get("maxLength").asInt()) {
                        throw DadoInvalidoException("Campo $nome é longo demais.")
                    }
                }
                "number", "integer" -> {
                    if (!valor.isNumber) {
                        throw DadoInvalidoException("Campo $nome deve ser numérico.")
                    }
                    if ("integer" == tipo && !valor.isIntegralNumber) {
                        throw DadoInvalidoException("Campo $nome deve ser inteiro.")
                    }
                }
                "boolean" -> {
                    if (!valor.isBoolean) {
                        throw DadoInvalidoException("Campo $nome deve ser verdadeiro ou falso.")
                    }
                }
                else -> {
                    // tipos compostos ficam para versões futuras do motor
                }
            }
            if (schema.has("enum")) {
                var bate = false
                for (opcao in schema.get("enum")) {
                    if (opcao == valor) {
                        bate = true
                        break
                    }
                }
                if (!bate) {
                    throw DadoInvalidoException("Campo $nome possui valor fora da lista permitida.")
                }
            }
        }
    }
}
